_drafts/partially-ordering-time.md

```diff
@@ -0,0 +1,253 @@
+up relaxing this strong view of time in order to gain speed.
+
+Weak memory models and consistency models are both _abstract_ models: they describe the
+_interface_ of the system to the programmer using it. They describe what sorts of
+behaviours we _can_ rely on from time, now that the default properties we tend to assume
+no longer apply.
+
+Both worlds have developed their own language for describing these models. For instance:
+the phrase “data race” in systems programming refers to any pair of accesses to a memory
+location that include at least one write, without a synchronisation barrier between
+them. This concept doesn't have a specific phrasing in distributed programming, because
+that field assumes that reads and writes participate in data races (“are concurrent”) by
+default.
+
+Thus, “data-race freedom” is a stronger condition than the distributed systems term
+“sequential consistency” (“any concurrent accesses can be treated _as if_ they were
+synchronised”), but if you can prove that a program on your processor is data-race free,
+you can _provide_ sequential consistency to that program as your hardware or software
+memory model (“SQ-DRF”, “sequential consistency for data-race free programs” [6]).
+
+…You can see how this could get confusing.
+
+The obvious solution is, of course, to make things more confusing!
+
+### The Burckhardt Paper
+
+The Burckhardt Paper. The _Burckhardt paper_.
+
+Dr. Sebastian Burckhardt has [been trying](#urls) to formalise the specification of
+consistency models for some time, in the distributed systems world. In [7], Appendix D.4,
+their team lays out an analysis of the C++ weak memory model in a slight specialisation of their framework.
+
+This gives us a convenient way to discuss the similarities and differences between the two
+fields. In particular, the C++ memory model _approximates_ to per-object causal
+consistency, with a few interesting differences.
+
+We'll study them axiom by axiom, but it's first important to note that where causal
+consistency is defined in (Burckhardt's) terms of a “visibility” and an “arbitration”
+orderings, they define weak memory axioms in terms of a “reads-from” and “modification
+order” instead, correspondingly.
+
+“Visibility” in the distributed systems model is a partial order of potential causality;
+it tracks what events (on possibly other replicas) are visible to what other events. It
+can contain read→read, read→write, write→read, and write→write edges, across multiple
+objects. “Reads-from” is much more specialised; it only ever contains write→read edges on
+the same object, and only ever contains one edge for a given read event.
+
+This corresponds to the fact that there only is a single actual memory cell being written
+to in a shared memory processor, for any given object, even if threads may access it at
+different points in causality. In a distributed system, a logical object can be written to
+at many separate locations (“replicas”).
```

Comment:
but each replica needs to be updated at some point right? if you accommodate replicas in the idea of an object, then can you use the "reads-from" edge in the same way in distributed systems?

Reply:
I'm not sure what your question here is. The distinction I'm trying to draw is how if you took a snapshot of a processor's memory, and asked "what should `read 0xcada12d0` return", there would only ever be one answer, regardless of which thread is asking. The same thing is not true of a distributed system, because the answer _does_ depend on which client/replica is asking.